sunpy/map/mapbase.py

```diff
@@ -1701,12 +1700,13 @@ def rotate(self, angle: u.deg = None, rmatrix=None, order=3, scale=1.0,
         unpad_x = -np.min((diff[0], 0))
         unpad_y = -np.min((diff[1], 0))
 
-        # Numpy 1.20+ prevents np.pad from padding with NaNs in integer arrays
-        # Before it would be cast to 0, but now it raises an error.
-        if version.parse(np.__version__) < version.parse("1.20.0") and issubclass(self.data.dtype.type, numbers.Integral) and (missing % 1 != 0):
-            warn_user("The specified `missing` value is not an integer, but the data "
-                      "array is of integer type, so the output may be strange.")
-        # Pad the image array
+        # Raise an informative error message if trying to pad an integer array with NaNs
+        if (pad_x > 0 or pad_y > 0 or missing % 1 != 0) and issubclass(self.data.dtype.type, numbers.Integral):
```

Comment:
```suggestion
        if (pad_x > 0 or pad_y > 0) and issubclass(self.data.dtype.type, numbers.Integral) and (missing % 1 != 0):
```
The change you made in adc72e0 has incorrect logic.  The error is now getting raised in situations where it shouldn't be raised.